nixos/installer: add kexec-boot

flokli opened the pull request:
This module exposes a config.system.build.kexecBoot attribute,
which returns a directory with kernel, initrd and a shell script
running the necessary kexec commands.

It's meant to be scp'ed to a machine with working ssh and kexec binary
installed.

This is useful for (cloud) providers where you can't boot a custom image, but
get some Debian or Ubuntu installation.

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

flokli committed:
nixos/installer: add kexec-boot

This module exposes a config.system.build.kexecBoot attribute,
which returns a directory with kernel, initrd and a shell script
running the necessary kexec commands.

It's meant to be scp'ed to a machine with working ssh and kexec binary
installed.

This is useful for (cloud) providers where you can't boot a custom image, but
get some Debian or Ubuntu installation.

Mic92 commented on nixos/modules/installer/kexec/kexec-boot.nix:
```diff
@@ -0,0 +1,50 @@
+}:
+{
+  imports = [
+    (modulesPath + "/installer/netboot/netboot-minimal.nix")
```
Just for reference. We could a similar approach like in build-vm: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/build-vm.nix#L12
to make the system.build.kexecBoot available by default without having to import a profile.

flokli replied:
This imports a configuration.nix (`qemu-vm.nix`) as a "base config", while we just use `config.system.build.toplevel` to have this compose with the current config. Can we pass a `config.system.build` to `extendModules`? (cc @roberth)

roberth replied:
`extendModules` creates a clone of the current configuration, similar to `specialisations` with `inheritParentConfig = true`, except the clone isn't automatically wired into anything; it's just a value.

`nixos-rebuild build-vm` needs this because you're not booting your regular config, but rather an adaptation of it.

In the case of `kexecBoot`, this looks like a way of booting what you intend to be the regular config.
I don't see what would be the benefit of cloning the entire config and what changes you would make to it. It does have a performance cost, evaluating all modules twice.

Mic92 replied:
However we don't want to include `(modulesPath + "/installer/netboot/netboot-minimal.nix")` by default but we want to provide `system.build.kexecBoot` by default.

roberth replied:
> we don't want to include `(modulesPath + "/installer/netboot/netboot-minimal.nix")` by default

That's right. We can't, because it's essentially a profile module outside the `profiles` directory. Profile-ness spreads to anything that imports it, as such modules contain not just options but configs that are enabled by default.


> but we want to provide `system.build.kexecBoot` by default.

If you can write the important parts of the profile as a module, you can have `system.build.kexecBoot` without `imports`-ing anything, and only a very minor, generally accepted performance impact.

Do you have a real world use case for generating two very similar configurations where you are both netbooting and booting in some other way, with minimal differences between the two?

flokli replied:
I think it's probably easier to keep this a separately-to-be-imported module for now, instead of always exposing a `kexecBoot` attribute. It's pretty consistent with what else lives in these folders.

Mic92 replied:
Ok. Eventually I might add `nixos-rebuild kexec`, if I get to it (tm) :)

roberth replied:
Unresolved for visibility. This is discussion is still relevant.

roberth replied:
I've been somewhat critical of this because of eval performance, but the simplicity of such an interface is very much worth it for use cases where the kexec variant isn't evaluated often. https://github.com/NixOS/rfcs/pull/22 would make it a no-brainer.

I think we could add such an interface by reusing applying the both modules in the kexec variant.
 - kexec-only use cases can keep using the current method
 - use cases where the kexec variant is only evaluated once benefit from the improved UX
 - use cases where both the kexec variant and the regular config are always needed simultaneously in the same evaluator invocation also benefit from the improved UX
 - the documentation can be simplified. Improving the performance of kexec-only invocations can be a side note with 1 example that most users can skip

roberth commented:
Could you add a VM test?

flokli committed:
nixosTests.kexec: extend with kexecBoot attribute

Add a node2, which imports the kexec-boot.nix profile.

Ensure node2 successfully boots up, then invoke the kexec-boot script
from it on node1.

flokli committed:
nixos/kexec-boot: use dirname of script to resolve bzImage and initrd.gz

This will allow invoking the `kexec-boot` script without `cd`-ing into
its folder first.

flokli committed:
nixosTests.kexec: fix tests with kexecBoot format

Lassulus commented:
so i debugged a bit. Problem was extracting the initrd failing with to much memory for me. Booting via systemd-boot inside the test vm fixed that. But now the backdoor seems to be unable to connect after kexec

Lassulus commented:
alright I pushed some changes, tests are now running fine for me. From my side this should be ready for merging.

roberth commented on nixos/tests/kexec.nix:
```diff
@@ -1,22 +1,46 @@
+    # Kexec node1 to the toplevel of node2 via the kexec-boot script
+    node1.succeed('touch /run/foo')
+    node1.execute('${nodes.node2.config.system.build.kexecBoot}/kexec-boot', check_return=False)
+    node1.succeed('! test -e /run/foo')
```
Maybe check the hostname or something else that identifies the `node2` config?
This will succeed erroneously if it just reboots.

Lassulus replied:
we should probably install a package like hello in node2 and see if it exists after kexec

roberth commented on nixos/tests/kexec.nix:
```diff
@@ -1,22 +1,46 @@
 # Test whether fast reboots via kexec work.
```
_fast_ isn't tested (nitpick)

Lassulus replied:
what else should we write?

roberth replied:
just remove _fast_, that works

```
# Test whether reboots via kexec work.
```

roberth replied:
It's actually a description of the first part of the test script now, which doesn't have a header comment.

roberth commented on nixos/modules/installer/kexec/kexec-boot.nix:
```diff
@@ -0,0 +1,51 @@
+  config = {
+    system.build.kexecBoot =
+      let
+        kexecScript = pkgs.writeScript "kexec-boot" ''
```
Maybe add:

```nix
# this script intentionally uses system dependencies for efficiency,
# instead of e.g. ${bash}.
```

Would be cool to have a script that rolls everything up into a tarball or into a self-unpacking script, so this can deployed by copying one file or an `ssh foo 'curl | sh'` one-liner.
Portability would also be a reason not to hardcode those deps.

roberth replied:
* portability to systems without a Nix installation

Lassulus replied:
maybe we can take some inspiration here: https://github.com/nix-community/nixos-generators/blob/master/formats/kexec.nix#L23

roberth commented:
We forgot to document kexec-boot and now this happens: https://github.com/NixOS/nixpkgs/pull/176004

flokli commented:
I'll write some docs.
